packages/mui-system/src/preprocessStyles.ts

```diff
@@ -0,0 +1,29 @@
+import { internal_serializeStyles } from '@mui/styled-engine';
+
+export default function preprocessStyles(styles: any) {
+  const variants = styles.variants
+
+  // Avoid passing `style.variants` to emotion, it will pollute the styles.
+  styles.variants = undefined
+  const serialized = internal_serializeStyles(styles) as any
+  styles.variants = variants
+
+  // Not supported on styled-components
+  if (serialized === styles) {
+    return styles;
+  }
```

Comment:
Is my understanding is correct that this PR only tackles Emotion, not included styled-components.

Reply:
Yes, I tried looking a bit into it but I'm not familiar with their codebase, not sure if the same thing is possible. They need to expose their serialization code and they need to accept pre-serialized objects as input for this optimization to make sense.

Reply:
I don't think it's important to fix styled-components's usage it's less than 10% of the usage: https://npm-stat.com/charts.html?package=%40mui%2Fstyled-engine&package=%40mui%2Fstyled-engine-sc&from=2023-08-29&to=2024-08-29